// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: user/v1/login.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on LoginRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *LoginRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LoginRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in LoginRequestMultiError, or
// nil if none found.
func (m *LoginRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *LoginRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if _, ok := Type_name[int32(m.GetType())]; !ok {
		err := LoginRequestValidationError{
			field:  "Type",
			reason: "value must be one of the defined enum values",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if _, ok := Method_name[int32(m.GetMethod())]; !ok {
		err := LoginRequestValidationError{
			field:  "Method",
			reason: "value must be one of the defined enum values",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetUnique()) < 1 {
		err := LoginRequestValidationError{
			field:  "Unique",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetSecret()) < 1 {
		err := LoginRequestValidationError{
			field:  "Secret",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return LoginRequestMultiError(errors)
	}

	return nil
}

// LoginRequestMultiError is an error wrapping multiple validation errors
// returned by LoginRequest.ValidateAll() if the designated constraints aren't met.
type LoginRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LoginRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LoginRequestMultiError) AllErrors() []error { return m }

// LoginRequestValidationError is the validation error returned by
// LoginRequest.Validate if the designated constraints aren't met.
type LoginRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LoginRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LoginRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LoginRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LoginRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LoginRequestValidationError) ErrorName() string { return "LoginRequestValidationError" }

// Error satisfies the builtin error interface
func (e LoginRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLoginRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LoginRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LoginRequestValidationError{}

// Validate checks the field values on OAuthLoginReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *OAuthLoginReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuthLoginReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// OAuthLoginReplyMultiError, or nil if none found.
func (m *OAuthLoginReply) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuthLoginReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for TokenType

	// no validation rules for ExpiresIn

	// no validation rules for Scope

	// no validation rules for AccessToken

	if m.RefreshToken != nil {
		// no validation rules for RefreshToken
	}

	if m.IdToken != nil {
		// no validation rules for IdToken
	}

	if len(errors) > 0 {
		return OAuthLoginReplyMultiError(errors)
	}

	return nil
}

// OAuthLoginReplyMultiError is an error wrapping multiple validation errors
// returned by OAuthLoginReply.ValidateAll() if the designated constraints
// aren't met.
type OAuthLoginReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuthLoginReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuthLoginReplyMultiError) AllErrors() []error { return m }

// OAuthLoginReplyValidationError is the validation error returned by
// OAuthLoginReply.Validate if the designated constraints aren't met.
type OAuthLoginReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuthLoginReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuthLoginReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuthLoginReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuthLoginReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuthLoginReplyValidationError) ErrorName() string { return "OAuthLoginReplyValidationError" }

// Error satisfies the builtin error interface
func (e OAuthLoginReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuthLoginReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuthLoginReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuthLoginReplyValidationError{}

// Validate checks the field values on LoginReply with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *LoginReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LoginReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in LoginReplyMultiError, or
// nil if none found.
func (m *LoginReply) ValidateAll() error {
	return m.validate(true)
}

func (m *LoginReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Token

	if len(errors) > 0 {
		return LoginReplyMultiError(errors)
	}

	return nil
}

// LoginReplyMultiError is an error wrapping multiple validation errors
// returned by LoginReply.ValidateAll() if the designated constraints aren't met.
type LoginReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LoginReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LoginReplyMultiError) AllErrors() []error { return m }

// LoginReplyValidationError is the validation error returned by
// LoginReply.Validate if the designated constraints aren't met.
type LoginReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LoginReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LoginReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LoginReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LoginReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LoginReplyValidationError) ErrorName() string { return "LoginReplyValidationError" }

// Error satisfies the builtin error interface
func (e LoginReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLoginReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LoginReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LoginReplyValidationError{}

// Validate checks the field values on LoginPrePhoneRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *LoginPrePhoneRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LoginPrePhoneRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// LoginPrePhoneRequestMultiError, or nil if none found.
func (m *LoginPrePhoneRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *LoginPrePhoneRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if !_LoginPrePhoneRequest_Phone_Pattern.MatchString(m.GetPhone()) {
		err := LoginPrePhoneRequestValidationError{
			field:  "Phone",
			reason: "value does not match regex pattern \"^\\\\+[1-9]\\\\d{1,14}$\"",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return LoginPrePhoneRequestMultiError(errors)
	}

	return nil
}

// LoginPrePhoneRequestMultiError is an error wrapping multiple validation
// errors returned by LoginPrePhoneRequest.ValidateAll() if the designated
// constraints aren't met.
type LoginPrePhoneRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LoginPrePhoneRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LoginPrePhoneRequestMultiError) AllErrors() []error { return m }

// LoginPrePhoneRequestValidationError is the validation error returned by
// LoginPrePhoneRequest.Validate if the designated constraints aren't met.
type LoginPrePhoneRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LoginPrePhoneRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LoginPrePhoneRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LoginPrePhoneRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LoginPrePhoneRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LoginPrePhoneRequestValidationError) ErrorName() string {
	return "LoginPrePhoneRequestValidationError"
}

// Error satisfies the builtin error interface
func (e LoginPrePhoneRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLoginPrePhoneRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LoginPrePhoneRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LoginPrePhoneRequestValidationError{}

var _LoginPrePhoneRequest_Phone_Pattern = regexp.MustCompile("^\\+[1-9]\\d{1,14}$")

// Validate checks the field values on LoginPreEmailRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *LoginPreEmailRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LoginPreEmailRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// LoginPreEmailRequestMultiError, or nil if none found.
func (m *LoginPreEmailRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *LoginPreEmailRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if err := m._validateEmail(m.GetEmail()); err != nil {
		err = LoginPreEmailRequestValidationError{
			field:  "Email",
			reason: "value must be a valid email address",
			cause:  err,
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return LoginPreEmailRequestMultiError(errors)
	}

	return nil
}

func (m *LoginPreEmailRequest) _validateHostname(host string) error {
	s := strings.ToLower(strings.TrimSuffix(host, "."))

	if len(host) > 253 {
		return errors.New("hostname cannot exceed 253 characters")
	}

	for _, part := range strings.Split(s, ".") {
		if l := len(part); l == 0 || l > 63 {
			return errors.New("hostname part must be non-empty and cannot exceed 63 characters")
		}

		if part[0] == '-' {
			return errors.New("hostname parts cannot begin with hyphens")
		}

		if part[len(part)-1] == '-' {
			return errors.New("hostname parts cannot end with hyphens")
		}

		for _, r := range part {
			if (r < 'a' || r > 'z') && (r < '0' || r > '9') && r != '-' {
				return fmt.Errorf("hostname parts can only contain alphanumeric characters or hyphens, got %q", string(r))
			}
		}
	}

	return nil
}

func (m *LoginPreEmailRequest) _validateEmail(addr string) error {
	a, err := mail.ParseAddress(addr)
	if err != nil {
		return err
	}
	addr = a.Address

	if len(addr) > 254 {
		return errors.New("email addresses cannot exceed 254 characters")
	}

	parts := strings.SplitN(addr, "@", 2)

	if len(parts[0]) > 64 {
		return errors.New("email address local phrase cannot exceed 64 characters")
	}

	return m._validateHostname(parts[1])
}

// LoginPreEmailRequestMultiError is an error wrapping multiple validation
// errors returned by LoginPreEmailRequest.ValidateAll() if the designated
// constraints aren't met.
type LoginPreEmailRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LoginPreEmailRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LoginPreEmailRequestMultiError) AllErrors() []error { return m }

// LoginPreEmailRequestValidationError is the validation error returned by
// LoginPreEmailRequest.Validate if the designated constraints aren't met.
type LoginPreEmailRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LoginPreEmailRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LoginPreEmailRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LoginPreEmailRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LoginPreEmailRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LoginPreEmailRequestValidationError) ErrorName() string {
	return "LoginPreEmailRequestValidationError"
}

// Error satisfies the builtin error interface
func (e LoginPreEmailRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLoginPreEmailRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LoginPreEmailRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LoginPreEmailRequestValidationError{}

// Validate checks the field values on TrashReply with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TrashReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TrashReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in TrashReplyMultiError, or
// nil if none found.
func (m *TrashReply) ValidateAll() error {
	return m.validate(true)
}

func (m *TrashReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for AuthToken

	// no validation rules for UserToken

	// no validation rules for ClientToken

	// no validation rules for Sub

	if len(errors) > 0 {
		return TrashReplyMultiError(errors)
	}

	return nil
}

// TrashReplyMultiError is an error wrapping multiple validation errors
// returned by TrashReply.ValidateAll() if the designated constraints aren't met.
type TrashReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TrashReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TrashReplyMultiError) AllErrors() []error { return m }

// TrashReplyValidationError is the validation error returned by
// TrashReply.Validate if the designated constraints aren't met.
type TrashReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TrashReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TrashReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TrashReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TrashReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TrashReplyValidationError) ErrorName() string { return "TrashReplyValidationError" }

// Error satisfies the builtin error interface
func (e TrashReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTrashReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TrashReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TrashReplyValidationError{}
